essence prime backend tool (DRAFT)

ewilbert7 opened the pull request:
In the draft phase of building the Essence ' backend tool.

**Updates**
- Set up structs
- "Essence Solver" set up to use custom temporary_directory to store intermediate files during solving process
- Made clear "TODO" notes for methods yet to be implemented

ozgurakgun commented on solvers/sr/src/essence_solver.rs:
```diff
@@ -0,0 +1,24 @@
+use std::path::{Path, PathBuf};
+use rand::Rng;
+
+pub struct EssenceSolver {
```
EssenceSolver isn't a good name for this, I'd say. It's either EssencePrimeSolver or SavileRowAsSolver maybe? But I think the files are at the right place and you will very quickly beat @niklasdewally's minion backend in terms of completeness :)

niklasdewally commented:
> just to note: this is as per #419
> 
> @niklasdewally did you manage to take a look at @ewilbert7's code? you set up the solver adaptor abstraction, would be good to get your feedback.


You should put your solver adaptor code somewhere in `conjure_core::solver::adaptors` not a seperate crate, otherwise you will run into a circular dependency when you later come to import this module in `conjure_core`. 

We only did solvers as separate crates for Minion / Chuffed because getting their apis into rust from C took a lot of lines of code and writing Rust bindings - if you don't need to do this, you can stick it all in core :) 

Apart from that LGTM so far!

ozgurakgun commented:
[This](https://github.com/conjure-cp/conjure-oxide/blob/85b2e45b6163cb4feac4a462695379541f60f70d/conjure_oxide/src/utils/conjure.rs#L109) is how we call Conjure to get all solutions.

ewilbert7 commented:
## Latest Commit Updates:

- Deleted separate sr_as_solver.rs file as it's unnecessary
- Implemented basic logic for solve, load_model and have implemented basic functions - get_family, get_name. Yet to implement solve_mut (not needed as of now)
- SavileRow struct now takes in the path to the Essence Prime file and the Solutions directory in the constructor
- Included comments of the idea of each code block with **questions/flags** for parts I need explanations/help with @ozgurakgun @niklasdewally (specifically, if that's the right way to call savilerow, solver stats)
- Added a helper functions to break down process, specifically **transform_to_essence_prime**. Just a placeholder for now until basic test runs are working.

ewilbert7 commented:
## Feedback 27th Nov

-  don't worry about transforming essence
- read and write from files
- run the tests already in the codebase
- download savilerow and test

niklasdewally commented:
FYI: I have since implemented some debugging pretty printing stuff that may be helpful here.

* Printing the model as essence in https://github.com/conjure-cp/conjure-oxide/pull/582
* Some miscellanious pretty printing functions  in `conjure_core::ast::pretty`
